crates/codegen/parser/runtime/src/napi_interface/cst.rs

```diff
@@ -68,10 +68,45 @@ impl RuleNode {
             .into()
     }
 
+    #[napi(catch_unwind, js_name = "toJSON")]
+    /// Serialize the token node to JSON.
+    ///
+    /// This method is intended for debugging purposes and may not be stable.
+    pub fn to_json(&self) -> String {
+        serde_json::to_string(&self.0).unwrap()
+    }
+
     #[napi(catch_unwind)]
     pub fn unparse(&self) -> String {
         self.0.clone().unparse()
     }
+
+    // Expose the children as a hidden (non-enumerable, don't generate type definition)
+    // property that's eagerly evaluated (getter) in the debugger context.
+    #[napi(
+        enumerable = false,
+        configurable = false,
+        writable = false,
+        getter,
+        js_name = "__children", // Needed; otherwise, the property name would shadow `children`.
+        skip_typescript
+    )]
+    pub fn __children(&self, env: Env) -> Vec<JsObject> {
+        Self::children(self, env)
+    }
+
+    // Similarly, expose the eagerly evaluated unparsed text in the debugger context.
+    #[napi(
```

Comment:
thinking out loud, if we are eagerly evaluating `__text()` recursively, does this mean generating the texts for all children nodes as well?
Would this blow up the debugger memory for large contracts?
The alternative is to just skip it, and have users call `unparse()` when needed.

Reply:
I dug a bit deeper to better understand who's responsible for what in this case and I believe we're safe in this scenario.

The core of the functionality here is the DAP `variables` request, which fetches *children* (not going deeper) for a given variable (container). From what I understand by going through vscode-js-debug, the extension provides a DAP adapter for the Chrome DevTools protocol exposed by Node when debugging. This DAP is called directly by VS Code to render the properties.

I will try to give a view into how the data flows but please excuse me for not making it crystal clear, I didn't want to spend too much time already into presenting all of the VS Code internals :sweat_smile: 
### DAP/DevTools Protocol

We have `GetterVariable` defined in https://github.com/microsoft/vscode-js-debug/blob/dfceaf103ce0cb83b53f1c3d88c06b8b63cb17da/src/adapter/variableStore.ts#L988.

The most important thing is that it has a `lazy: true` presentationHint in `toDap`.
  
  It's created via `createPropertyVar`, which is called by `createObjectPropertyVars`. Nothing here is recursive but it exposes a `getChildren()`, which remotely calls the getter via `Runtime` (Dev Tools) if children are requested.

These properties are constructed for requested DAP `variables` in https://github.com/microsoft/vscode-js-debug/blob/dfceaf103ce0cb83b53f1c3d88c06b8b63cb17da/src/adapter/debugAdapter.ts#L362. Inside the called `getVariables`, we only fetch the children of a requested scope/variable container.

### VS Code

The DAP `variables` request is being sent in https://github.com/microsoft/vscode/blob/344682bd1717fabe4e71323760b63becca474633/src/vs/workbench/contrib/debug/common/debugModel.ts#L154 as part of the `ExpressionContainer`. It's called as part of [`(do)GetChildren()`](https://github.com/microsoft/vscode/blob/344682bd1717fabe4e71323760b63becca474633/src/vs/workbench/contrib/debug/common/debugModel.ts#L107).

What's crucial here is that `ExpressionContainer#hasChildren` will return false if it's lazy (`presentationHint?.lazy`): https://github.com/microsoft/vscode/blob/344682bd1717fabe4e71323760b63becca474633/src/vs/workbench/contrib/debug/common/debugModel.ts#L149.

The container itself is used as the shown data source for the trees in the different view panes.

Let's take the  watch expressions view pane as an example.

1. the initial tree is seeded with the `IDebugService` at https://github.com/microsoft/vscode/blob/344682bd1717fabe4e71323760b63becca474633/src/vs/workbench/contrib/debug/browser/watchExpressionsView.ts#L114
2. the tree is created at the first level with with [`WatchExpressionsDataSource`](https://github.com/microsoft/vscode/blob/344682bd1717fabe4e71323760b63becca474633/src/vs/workbench/contrib/debug/browser/watchExpressionsView.ts#L256), which seeds the initial list with `getWatchExpressions(): Expression[]`.
3. Now, since the `Expression extends ExpressionContainer`, it will return `hasChildren() = false` if it's a (lazy) getter.
4. As such, the lazy getters will not be evaluated recursively, and so this will not recursively blow up when inspecting/watching a CST node expression in the debugger.

## Takeaways

The getters are not evaluated recursively, so we won't evaluate the entire tree greedily and their unparsed text.

Moreover, our tree does not overlap in terms of the source code contents, so ignoring the overhead of single allocations, even if the user aggressively expanded the entire tree manually, the additional memory footprint would not take more memory in the debugger in the debugger as the entire node itself (accounting just for `__text`-related memory; I doubt we will have >100MB source worth of contracts).

As such, I think it's fine to leave the `__text` as-is and it will not cause memory problems for bigger contracts.

Reply:
To be also clear what `eager evaluation` means here, it's that it will be evaluated as if it was a state variable, on the child/property level for an object, rather than having to call/evaluate it manually each time (e.g. if it was a regular object method). As I explained above, it does not mean eager recursive evaluation.

Reply:
Wonderful analysis and explanation as usual! Thanks for double checking! :)